go/stellar/note.go

```diff
@@ -0,0 +1,236 @@
+	if err != nil {
+		return res, err
+	}
+	var recipient *stellar1.NoteRecipient
+	if other != nil && !other.Eq(me.ToUserVersion()) {
+		u2, err := loadUvUpk(ctx, g, *other)
+		if err != nil {
+			return res, fmt.Errorf("error loading recipient: %v", err)
+		}
+		puk2 := u2.GetLatestPerUserKey()
+		if puk2 == nil {
+			return res, fmt.Errorf("recipient has no per-user key")
+		}
+		// Overwite symmetricKey with the shared key.
+		symmetricKey, err = noteMixKeys(ctx, g, puk1Seed, puk2.EncKID)
+		if err != nil {
+			return res, err
+		}
+		recipient = &stellar1.NoteRecipient{
+			User:   *other,
+			PukGen: keybase1.PerUserKeyGeneration(puk2.Gen),
+		}
+	}
+	return noteBuildSecret{
+		symmetricKey: symmetricKey,
+		sender: stellar1.NoteRecipient{
+			User:   me.ToUserVersion(),
+			PukGen: puk1Gen,
+		},
+		recipient: recipient,
+	}, nil
+}
+
+func noteSymmetricKeyForDecryption(ctx context.Context, g *libkb.GlobalContext, encNote stellar1.EncryptedNote) (res libkb.NaclSecretBoxKey, err error) {
+	me, err := loadMeUpk(ctx, g)
+	if err != nil {
+		return res, err
+	}
+	var mePukGen keybase1.PerUserKeyGeneration
+	var them *stellar1.NoteRecipient
+	if encNote.Sender.User.Eq(me.ToUserVersion()) {
+		mePukGen = encNote.Sender.PukGen
+		them = encNote.Recipient
+	}
+	if encNote.Recipient != nil && encNote.Recipient.User.Eq(me.ToUserVersion()) {
+		mePukGen = encNote.Recipient.PukGen
+		them = &encNote.Sender
+	}
+	if mePukGen == 0 {
+		return res, fmt.Errorf("note not encrypted for logged-in user")
+	}
+	pukring, err := g.GetPerUserKeyring()
+	if err != nil {
+		return res, err
+	}
+	err = pukring.Sync(ctx)
+	if err != nil {
+		return res, err
+	}
+	pukSeed, err := pukring.GetSeedByGeneration(ctx, mePukGen)
+	if err != nil {
+		return res, err
+	}
+	if them == nil {
+		return pukSeed.DeriveSymmetricKey(libkb.DeriveReasonPUKStellarNoteSelf)
+	}
+	u2, err := loadUvUpk(ctx, g, them.User)
+	if err != nil {
+		return res, err
+	}
+	puk2 := u2.GetPerUserKeyByGen(them.PukGen)
+	if puk2 == nil {
+		return res, fmt.Errorf("could not find other user's key: %v %v", them.User.String(), them.PukGen)
+	}
+	return noteMixKeys(ctx, g, pukSeed, puk2.EncKID)
+}
+
+// noteMixKeys derives a shared symmetric key for two DH keys.
+// The key is the last 32 bytes of the nacl box of 32 zeros with a use-specific nonce.
+func noteMixKeys(ctx context.Context, g *libkb.GlobalContext, puk1 libkb.PerUserKeySeed, puk2EncKID keybase1.KID) (res libkb.NaclSecretBoxKey, err error) {
+	puk1Enc, err := puk1.DeriveDHKey()
+	if err != nil {
+		return res, err
+	}
+	if puk1Enc.Private == nil {
+		return res, fmt.Errorf("missing local puk private key")
+	}
+	puk2EncGeneric, err := libkb.ImportKeypairFromKID(puk2EncKID)
+	if err != nil {
+		return res, err
+	}
+	puk2Enc, ok := puk2EncGeneric.(libkb.NaclDHKeyPair)
+	if !ok {
+		return res, fmt.Errorf("recipient per-user key was not a DH key")
+	}
+	var zeros [32]byte
+	nonce := noteMixPukNonce()
```

Comment:
Let's explain: This is a constant nonce used for key derivation. The keys that we spit out will then be used with real one-time-only nonces of their own for actual encryption/decryption.